### PR TITLE
When supplying GradientDescent with gradients, they implicitly define which parameters to update

### PR DESCRIPTION
**Pull request by jbornschein**
Currently the 'params' attribute of DifferentiableCostMinimizer is always initialized to contain all shared variables of the cost-graph; even when the user supplies an explicit gradients-dict to GradientDescent. 

1) I guess we can trust the user and automatically set it to gradients.keys()?
2) I wonder whether we should guard GradientDescent.gradients with a property/setter?

### CODE REVIEW

File: blocks/algorithms/__init__.py [hunk 1]
```diff
@@ -180,7 +180,11 @@ class GradientDescent(DifferentiableCostMinimizer):
 
     """
     def __init__(self, step_rule=None, gradients=None, **kwargs):
-        super(GradientDescent, self).__init__(**kwargs)
+        if gradients and 'params' not in kwargs:
```
Comment (bartvm) [inline blocks/algorithms/__init__.py]: I guess this should be `'gradients'`?

Comment (bartvm) [inline blocks/algorithms/__init__.py]: Oh, but wait, `gradients` is an argument, so it will never be in `kwargs`

Comment (jbornschein) [inline blocks/algorithms/__init__.py]: This indeed looks confusing -- maybe it should read:

`if gradients is not Null and 'params' not in kwargs:`

I was also wondering: Maybe DifferentiableCostMinimizer shouldn't populate `params` at all? Identifying parameter variables that can be optimized is maybe something a more specialized subclass should do? E.g. GradientDescent. That would simplify/streamline the code here a bit.

Comment (rizar) [inline blocks/algorithms/__init__.py]: Would this do the job?

```python
kwargs.setdefault("params", gradients.keys())
super(GradientDescent, self).__init__(**kwargs)
```

Comment (jbornschein) [inline blocks/algorithms/__init__.py]: if gradients:
        kwargs.setdefault("params", gradients.keys())
    super(GradientDescent, self).__init__(**kwargs)

would work and looks better than the current proposal.

Comment (rizar) [inline blocks/algorithms/__init__.py]: Cool, let's do that and merge then!

**Pull request merged**